### PR TITLE
[Enhancement] Support global runtime filters for nestloop join

### PR DESCRIPTION
**Pull request by JasonQu1**
## Why I'm doing:
StarRocks already supports local runtime filters for NLJoin when the build side can be reduced to runtime values, but those filters are limited to the local execution path and cannot be delivered to remote probe-side scan nodes.

## What I'm doing

This change adds global runtime filter support for NLJoin range predicates.

Main changes:

- Allow eligible NLJoin range runtime filters to be generated as global RF instead of always forcing them to local-only.
- Normalize NLJoin predicates so the probe-side expression is on the left and the build-side expression is on the right before RF generation.
- For range predicates, compute the build-side min/max boundary and use it to build a MinMax runtime filter.
- Publish NLJoin global RF through the existing runtime filter publish path.
- Keep the existing local RF behavior for single-row build cases.
- Add handling for NULL / NaN boundary values so invalid build-side values do not produce unsafe filters.
- Add FE plan tests, SQL regression tests, and BE unit tests for the new boundary computation logic.

### Environment

- Cluster: 1 FE + 1 BE
- Probe table: `fact_probe_indexed`, 100,000,010 rows
- Build table: `build_range_multi_indexed`, 3 rows

### Table Schema
```
CREATE TABLE fact_probe_indexed (
    event_day DATE NULL,
    id BIGINT NULL,
    payload VARCHAR(512) NULL,
    INDEX idx_event_day (event_day) USING BITMAP
);

CREATE TABLE build_range_multi_indexed (
    bound_day DATE NULL
);

INSERT INTO build_range_multi_indexed VALUES
    ('2024-03-10'),
    ('2024-03-20'),
    ('2024-03-25');
```
### Test SQL

```
SELECT sum(length(j.payload))
FROM (
    SELECT event_day, payload
    FROM fact_probe_indexed
    GROUP BY event_day, payload
) j
JOIN build_range_multi_indexed b
ON j.event_day > b.bound_day;
```

### Result

| Case | Avg Latency | Avg Peak Memory | ZoneMap filtered rows | RF Output |
| --- | ---: | ---: | ---: | ---: |
| Global RF off | 16.090s | 63.979 GB | 0 | 0 |
| Global RF on | 5.652s | 25.717 GB | 67.633M | 31.000M |

### Summary

- Average latency improved from `16.090s` to `5.652s`.
- Peak memory decreased from `63.979 GB` to `25.717 GB`.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.1
  - [ ] 4.0
  - [ ] 3.5

### CODE REVIEW

**Comment by JasonQu1** [comment]
@codex review

File: fe/fe-core/src/main/java/com/starrocks/planner/NestLoopJoinNode.java [hunk 1]
```diff
@@ -76,9 +77,12 @@ public void buildRuntimeFilters(IdGenerator<RuntimeFilterId> generator, Descript
                 rf.setExprOrder(i);
                 rf.setJoinMode(DistributionMode.BROADCAST);
                 rf.setBuildCardinality(buildStageNode.getCardinality());
-                rf.setOnlyLocal(true);
                 rf.setBuildExpr(right);
 
+                if (!(expr instanceof BinaryPredicate) || !BinaryPredicate.IS_RANGE_PREDICATE.apply((BinaryPredicate) expr)) {
+                    rf.setOnlyLocal(true);
```
[the remark this answers is not in the document]
Comment (JasonQu1) [reply fe/fe-core/src/main/java/com/starrocks/planner/NestLoopJoinNode.java:86]: JSON and VARIANT types cannot be used as join conditions. FE rejects them during analysis via Type.canJoinOn(), so this code path is unreachable.

**Comment by JasonQu1** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by JasonQu1** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by JasonQu1** [comment]
rerun

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by CelerData-Reviewer** [comment]
@codex review
